Proxy DinD calls of airbyte-ci through Tailscale to access pull through cache

cpdeethree opened the pull request:
## What

DinD needs to be able to see our internal pull through cache due to rate limiting issues. The pull through cache is currently referenced by a kubernetes service IP that is visible to things running inside the cluster. The Dagger engine can see it because it's accessible to the host network, and we inject the serviceIP of the pull through cache via the engine configmap in the daemonset, but there is no way in the current dagger version (0.6.4) for containers spawned by dagger such as the DinD container that runs testcontainers to access the host network. Well, actually, there is a way, but we won't do it, because it is totally unsupported and involves some serious hackity hackersons. Dagger 0.9.0+ does offer a supported way, so if we wanted to go that route instead this tailscale fun would not be necessary.


## How

This uses [userspace networking mode](https://tailscale.com/kb/1112/userspace-networking/) to run tailscale as a sidecar to the dind container/service. This will allow DinD calls to access the tailscale network that gives access to the vpc Another way this could theoretically be done is by running tailscale itself as a background process in the dind container, but that is actually just not possible with Dagger, it will always hang, it doesn't like background processes, it always wants to return an exit code. Trust me, I tried that first ;) 

So you need to sidecar it with a service like what I've done.

## TODO:

- [x] Pass TAILSCALE_AUTHKEY into this function, maybe only when we are in CI Currently this will always fail because nothing is creating this secret and passing it in
- [x] Hide bootstrapping the sidecar container when the TAILSCALE_AUTHKEY secret is not passed in, such as in some local run
- [x] Verify that the HTTP calls that docker makes to resolve images respect ALL_PROXY. They should, but we need to test to make sure.
- [x] Verify that this cluster IP is even visible via tailscale. We can probably just curl the endpoint to check. If it isn't, we can do like what we did with prefect.airbyte.com and create some arbitrary dns entry that resolves to an internal ip instead

cpdeethree commented on airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:
```diff
@@ -43,7 +63,7 @@ def with_global_dockerd_service(
                     # Update daemon config with storage driver.
                     "mkdir /etc/docker",
-                    '(echo {\\"storage-driver\\": \\"fuse-overlayfs\\"} > /etc/docker/daemon.json)',
+                    'echo \'{"storage-driver": "fuse-overlayfs", "registry-mirrors": ["http://172.20.83.84:5000"], "insecure-registries": ["http://172.20.83.84:5000"]}\' > /etc/docker/daemon.json',
```
I'm not sure if tailscale is gonna be able to see this service IP, i think it will, but if not then we will front this with the internal load balancer like we do with prefect

alafanechere commented:
@cpdeethree thank you for the stub.
I reworked the code a bit to make it more readable and optionally bind `dockerd` to `tailscale`.
So far I can confirm that the tailscale binding is working: `curl` to `prefect.airbyte.com` works.
But `dockerd` can reach the registry mirror (pinging `172.20.83.84:5000` does not work). 

Could you:
> front this with the internal load balancer like we do with prefect and make a cloudflare entry that points at an internal DNS entry, same as prefect.airbyte.com

alafanechere commented on .github/actions/run-dagger-pipeline/action.yml:
```diff
@@ -16,6 +20,11 @@ inputs:
+  docker_registry_mirror_url:
+    description: "Docker registry mirror URL"
+    required: false
+    # TODO (Conor): Change this value to the correct URL once we have a reachable mirror
```
cc @cpdeethree

alafanechere commented on .github/workflows/connectors_tests.yml:
```diff
@@ -20,6 +20,14 @@ on:
+      airbyte_ci_binary_url:
+        description: "The URL to the airbyte-ci binary to use"
+        default: "https://connectors.airbyte.com/airbyte-ci/releases/ubuntu/latest/airbyte-ci"
+      # TODO (Conor): Remove this once the default value is correctly set in ./.github/actions/run-dagger-pipeline
```
cc @cpdeethree

alafanechere commented on .github/workflows/connectors_tests.yml:
```diff
@@ -55,9 +63,12 @@ jobs:
           context: "manual"
           docker_hub_password: ${{ secrets.DOCKER_HUB_PASSWORD }}
           docker_hub_username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          # TODO (Conor): Remove this once the default value is correctly set in ./.github/actions/run-dagger-pipeline
```
cc @cpdeethree

alafanechere reviewed (changes requested):
@cpdeethree I don't think there's anything to change on the python side of things following my work.
The remaining things to do are:
* [x] Pass a reachable `docker_registry_mirror_url` [here](https://github.com/airbytehq/airbyte/pull/32965/files#diff-8bba74a834b43cc4f45b74d9e25f898d291bc2d8c537e423325e4da1487efaa1R27)
* [x] Confirm the `testcontainers` are pulled through cache

The current state of the code will lead to an execution failure if the mirror URL is not reachable.

You can test out any new URL for the mirror in [connector workflow input ](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml). 
You should pass the following binary url:
`https://storage.googleapis.com/dev-airbyte-cloud-connector-metadata-service/airbyte-ci/releases/ubuntu/42aa44a/airbyte-ci`

And try with `--name=source-postgres`
<img width="317" alt="Screenshot 2023-11-30 at 12 48 41" src="https://github.com/airbytehq/airbyte/assets/5551758/573193af-799c-4ba0-a249-82a19026fdf7">

cpdeethree commented:
doin stuff

https://github.com/airbytehq/airbyte/actions/runs/7051101868

![Screenshot 2023-11-30 at 1 05 07 PM](https://github.com/airbytehq/airbyte/assets/10550441/16cf8ed5-b930-4312-b43e-8c55af53dafe)

alafanechere reviewed (changes requested):
I ran `source-postgres` test with the latest binary on this branch and faced an exception in integration tests using test containers:
```
org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=postgres:16-bullseye
```

[Test report](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/manual/dind-with-tailscale/1701428612/f198f8d99ed1ff62f0521dd2a2bd02f559447004/source-postgres/3.2.24/output.html)

cpdeethree commented:
This looks to be working now. I think the final step is to revert [13f9587](https://github.com/airbytehq/airbyte/pull/32965/commits/13f9587441cf316f01851742a596a279db8f7c19) and perhaps some final load testing if we want to be super confident

cpdeethree commented on airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:
```diff
@@ -8,60 +8,178 @@
     )
-    if docker_hub_username_secret and docker_hub_password_secret:
-        dockerd_container = (
+    if docker_var_lib_caching:
+        # We cache /var/lib/docker to avoid downloading images and layers multiple times.
+        base_container = base_container.with_mounted_cache("/var/lib/docker", dagger_client.cache_volume(DOCKER_VAR_LIB_VOLUME_NAME))
```
I got a passing run after you did that here:https://github.com/airbytehq/airbyte/actions/runs/7065507194

stephane-airbyte commented:
Current dependencies on/for this PR:
* master
  * **PR #32965** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/32965?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #33034** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33034?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/32965?utm_source=stack-comment).

cpdeethree commented:
GCS Bucket containing pull through cache registry entries before I changed the source-postgres testcontainers java to target bookworm instead of bullseye

![Screenshot 2023-12-06 at 11 20 54 AM](https://github.com/airbytehq/airbyte/assets/10550441/b19aca4e-3c1a-4f5f-83d4-ba854d023f2a)

GCS Bucket containing pull through cache registry entries after I changed the soruce-postgres testcontainers java to target bookworm instead of bullseye

![Screenshot 2023-12-06 at 11 27 46 AM](https://github.com/airbytehq/airbyte/assets/10550441/ea137e0f-fbd2-464d-9b81-b531c601e8c7)

postamar commented on airbyte-ci/connectors/pipelines/pipelines/consts.py:
```diff
@@ -53,6 +56,8 @@
 PIP_CACHE_PATH = "/root/.cache/pip"
 POETRY_CACHE_VOLUME_NAME = "poetry_cache"
 POETRY_CACHE_PATH = "/root/.cache/pypoetry"
+STORAGE_DRIVER = "fuse-overlayfs"
```
I'm not sure that this constant makes sense. It's only a coincidence that its alpine package name is the same string as the value in the dockerd config.

cpdeethree replied:
This is actually something that augustin added, so I'm a bit loath to touch it, but maybe we can fix this in a followup pr cc @alafanechere

postamar commented on airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:
```diff
@@ -8,62 +8,145 @@
-) -> Container:
-    """Create a container with a docker daemon running.
-    We expose its 2375 port to use it as a docker host for docker-in-docker use cases.
+def get_base_dockerd_container(dagger_client: Client, docker_var_lib_caching: bool = True) -> Container:
```
Why do we still have this argument? Also the docstring text doesn't match, regarding the default value. Just remove it, it's always going to be always-on or always-off at all callsites.

cpdeethree replied:
I'l;l just change this back to always on, like you said

postamar commented on airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:
```diff
@@ -8,62 +8,145 @@
+    dockerd_container: Container,
+    docker_registry_username_secret: Optional[Secret],
+    docker_registry_password_secret: Optional[Secret],
+    docker_registry_address: Optional[str] = "docker.io",
```
Similar comment regarding function arg that doesn't really serve a purpose. Here a constant would be better.

cpdeethree replied:
moved to a const

postamar commented on airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:
```diff
@@ -8,62 +8,145 @@
+        daemon_config_json = get_daemon_config_json()
+
+    dockerd_container = dockerd_container.with_new_file("/etc/docker/daemon.json", daemon_config_json)
+    # Docker login happens late because there's a cache buster in the docker login command.
```
Actually it would probably be fine to have the cachebuster present in all cases, perhaps just after line 134. We wouldn't pay that much of a performance penalty I don't think, and it might simplify things a bit.

cpdeethree replied:
we already bust this above a few lines before too. I think in practice like you said the difference is almost completely negligble because of that

postamar commented on airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:
```diff
@@ -8,62 +8,145 @@
+    dockerd_container = dockerd_container.with_new_file("/etc/docker/daemon.json", daemon_config_json)
+    # Docker login happens late because there's a cache buster in the docker login command.
+    dockerd_container = docker_login(dockerd_container, docker_hub_username_secret, docker_hub_password_secret)
+    return dockerd_container.with_exec(  # TODO: Remove this dummy config after testing
```
either handle TODO or remove it?

postamar reviewed (approved):
This is marvellous, thanks for your good work! I only have minor comments, which should be straightforward to address prior to merging.

I'm unable to see what makes this work vs. an earlier version. Is this a case of getting the magic incantation _just right_?
